Build failures with current tree

We have the following build failures:

```
warp7_m4                  tests/lib/sprintf/libraries.libc                   FAILED: build_error
        see: sanity-out/warp7_m4/tests/lib/sprintf/libraries.libc/build.log

total complete:    1/  11   9%  failed:    1

96b_meerkat96             tests/lib/sprintf/libraries.libc                   FAILED: build_error
        see: sanity-out/96b_meerkat96/tests/lib/sprintf/libraries.libc/build.log

total complete:    2/  11  18%  failed:    2

colibri_imx7d_m4          tests/lib/sprintf/libraries.libc                   FAILED: build_error
        see: sanity-out/colibri_imx7d_m4/tests/lib/sprintf/libraries.libc/build.log

total complete:    3/  11  27%  failed:    3

rv32m1_vega_ri5cy         samples/subsys/logging/logger/samples.logger       FAILED: build_error
        see: sanity-out/rv32m1_vega_ri5cy/samples/subsys/logging/logger/samples.logger/build.log

total complete:    4/  11  36%  failed:    4

bbc_microbit              samples/bluetooth/mesh_demo/sample.bluetooth.mesh_demo FAILED: build_error
        see: sanity-out/bbc_microbit/samples/bluetooth/mesh_demo/sample.bluetooth.mesh_demo/build.log

total complete:    5/  11  45%  failed:    5

nucleo_l4r5zi             samples/subsys/nvs/sample.subsys.nvs               FAILED: build_error
        see: sanity-out/nucleo_l4r5zi/samples/subsys/nvs/sample.subsys.nvs/build.log

total complete:    6/  11  54%  failed:    6

bbc_microbit              samples/bluetooth/hci_uart/sample.bluetooth.hci_uart.arm FAILED: build_error
        see: sanity-out/bbc_microbit/samples/bluetooth/hci_uart/sample.bluetooth.hci_uart.arm/build.log

total complete:    9/  11  81%  failed:    7

nrf52810_pca10040         tests/kernel/mutex/sys_mutex/sys.mutex             FAILED: build_error
        see: sanity-out/nrf52810_pca10040/tests/kernel/mutex/sys_mutex/sys.mutex/build.log

total complete:   10/  11  90%  failed:    8

nrf52811_pca10056         tests/kernel/mutex/sys_mutex/sys.mutex             FAILED: build_error
        see: sanity-out/nrf52811_pca10056/tests/kernel/mutex/sys_mutex/sys.mutex/build.log
```

The majority of these are code or memory size issues.

The rv32m1_vega_ri5cy is reported in #18198

the nucleo_l4r5zi is a compile error:

```
/home/galak/git/FULL_ZEPHYR/zephyr/drivers/flash/flash_stm32l4x.c: In function 'flash_stm32_block_erase_loop':
/home/galak/git/FULL_ZEPHYR/zephyr/drivers/flash/flash_stm32l4x.c:149:11: error: 'pages_per_bank' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  if (page >= pages_per_bank)
      ~~~~~^~~~~~~~~~~~~~~~~
/home/galak/git/FULL_ZEPHYR/zephyr/drivers/flash/flash_stm32l4x.c:118:8: note: 'pages_per_bank' was declared here
  u16_t pages_per_bank;
        ^~~~~~~~~~~~~~
```

--- a/drivers/flash/flash_stm32l4x.c
+++ b/drivers/flash/flash_stm32l4x.c
@@ -49,9 +49,9 @@ static int write_dword(struct device *dev, off_t offset, u64_t val)
 {
 	volatile u32_t *flash = (u32_t *)(offset + CONFIG_FLASH_BASE_ADDRESS);
 	struct stm32l4x_flash *regs = FLASH_STM32_REGS(dev);
-#ifdef FLASH_OPTR_DUALBANK
+#if defined(FLASH_OPTR_DUALBANK) || defined(FLASH_OPTR_DBANK)
 	bool dcache_enabled = false;
-#endif /* FLASH_OPTR_DUALBANK */
+#endif /* FLASH_OPTR_DUALBANK || FLASH_OPTR_DBANK */
 	u32_t tmp;
 	int rc;
 
@@ -72,7 +72,7 @@ static int write_dword(struct device *dev, off_t offset, u64_t val)
 		return -EIO;
 	}
 
-#ifdef FLASH_OPTR_DUALBANK
+#if defined(FLASH_OPTR_DUALBANK) || defined(FLASH_OPTR_DBANK)
 	/*
 	 * Disable the data cache to avoid the silicon errata 2.2.3:
 	 * "Data cache might be corrupted during Flash memory read-while-write operation"
@@ -81,7 +81,7 @@ static int write_dword(struct device *dev, off_t offset, u64_t val)
 		dcache_enabled = true;
 		regs->acr.val &= (~FLASH_ACR_DCEN);
 	}
-#endif /* FLASH_OPTR_DUALBANK */
+#endif /* FLASH_OPTR_DUALBANK || FLASH_OPTR_DBANK */
 
 	/* Set the PG bit */
 	regs->cr |= FLASH_CR_PG;
@@ -99,14 +99,14 @@ static int write_dword(struct device *dev, off_t offset, u64_t val)
 	/* Clear the PG bit */
 	regs->cr &= (~FLASH_CR_PG);
 
-#ifdef FLASH_OPTR_DUALBANK
+#if defined(FLASH_OPTR_DUALBANK) || defined(FLASH_OPTR_DBANK)
 	/* Reset/enable the data cache if previously enabled */
 	if (dcache_enabled) {
 		regs->acr.val |= FLASH_ACR_DCRST;
 		regs->acr.val &= (~FLASH_ACR_DCRST);
 		regs->acr.val |= FLASH_ACR_DCEN;
 	}
-#endif /* FLASH_OPTR_DUALBANK */
+#endif /* FLASH_OPTR_DUALBANK || FLASH_OPTR_DBANK */
 
 	return rc;
 }
@@ -118,16 +118,30 @@ static int erase_page(struct device *dev, unsigned int page)
 	u16_t pages_per_bank;
 	int rc;
 
-#if DT_FLASH_SIZE == 1024
-	/* The SOC has 1MB of FLASH with DUALBANKing by design */
-	pages_per_bank = 256U;
-#elif DT_FLASH_SIZE < 1024
-#ifdef FLASH_OPTR_DUALBANK
-	/* DUALBANKing is used, so find the correct number of pages per bank */
-	pages_per_bank = DT_FLASH_SIZE >> 2; /* Also there are two banks */
-#else
-	pages_per_bank = DT_FLASH_SIZE >> 1; /* Each page is of 2KB size */
-#endif /* FLASH_OPTR_DUALBANK */
+#if !defined(FLASH_OPTR_DUALBANK) && !defined(FLASH_OPTR_DBANK)
+	/* Single bank device. Each page is of 2KB size */
+	pages_per_bank = DT_FLASH_SIZE >> 1;
+#elif defined(FLASH_OPTR_DUALBANK)
+	/* L4 series (2K page size) with configurable Dual Bank (default y) */
+	/* Dual Bank is only option for 1M devices */
+	if ((regs->optr & FLASH_OPTR_DUALBANK) || (DT_FLASH_SIZE == 1024)) {
+		/* Dual Bank configuration (nbr pages = flash size / 2 / 2K) */
+		pages_per_bank = DT_FLASH_SIZE >> 2;
+	} else {
+		/* Single bank configuration. This has not been validated. */
+		/* Not supported for now. */
+		return -ENOTSUP;
+	}
+#elif defined(FLASH_OPTR_DBANK)
+	/* L4+ series (4K page size) with configurable Dual Bank (default y)*/
+	if (regs->optr & FLASH_OPTR_DBANK) {
+		/* Dual Bank configuration (nbre pags = flash size / 2 / 4K) */
+		pages_per_bank = DT_FLASH_SIZE >> 3;
+	} else {
+		/* Single bank configuration */
+		/* Requires 128 bytes data read. This config is not supported */
+		return -ENOTSUP;
+	}
 #endif
 
 	/* if the control register is locked, do not fail silently */

